compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -663,6 +700,41 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                             return;
                         }
                         let trait_ref = trait_predicate.to_poly_trait_ref();
+
+                        // FIXME(bryangarza): Move this logic to its own function
+                        // When using the new trait solver, recompute the safe transmute reason and use that for the error.
+                        if Some(trait_ref.def_id()) == self.tcx.lang_items().transmute_trait()
+                            && self.tcx.trait_solver_next()
+                        {
+                            let cause = obligation.cause.clone();
+                            let src_and_dst =
+                                trait_predicate.map_bound(|p| rustc_transmute::Types {
+                                    dst: p.trait_ref.substs.type_at(0),
+                                    src: p.trait_ref.substs.type_at(1),
+                                });
+                            let scope = trait_ref.skip_binder().substs.type_at(2);
+                            let Some(assume) =
+                                rustc_transmute::Assume::from_const(self.infcx.tcx, obligation.param_env, trait_ref.skip_binder().substs.const_at(3)) else {
+                                    bug!("Unable to construct rustc_transmute::Assume where it was previously possible.");
+                                };
+                            match rustc_transmute::TransmuteTypeEnv::new(self.infcx)
+                                .is_transmutable(cause, src_and_dst, scope, assume)
+                            {
+                                // Should never get a Yes at this point! We already ran it before, and did not get a Yes.
+                                rustc_transmute::Answer::Yes => bug!(
+                                    "Inconsistent rustc_transmute::is_transmutable(...) result."
+                                ),
+                                rustc_transmute::Answer::No(reason) => {
+                                    return self.report_selection_error(
+                                        obligation,
+                                        root_obligation,
+                                        &SelectionError::NotTransmutable(reason),
+                                    );
+                                }
+                                _ => (),
+                            };
+                        }
```

Comment:
@compiler-errors Added this so the new trait solver can emit the right error, is this what you had in mind?

Reply:
I think I asked this before, but why do we even need a special selection error kind? You have shown that it's possible to recompute the transmutability reason on the fly here :)

Reply:
It was just to avoid computing the safe transmute reason twice, but if the new trait solver is going to need it 2x anyways, we could just remove `SelectionError::NotTransmutable` variant